clippy_lints/src/redundant_closure_call.rs

```diff
@@ -61,10 +61,15 @@ impl<'tcx> Visitor<'tcx> for ReturnVisitor {
 }
 
 /// Checks if the body is owned by an async closure
-fn is_async_closure(body: &hir::Body<'_>) -> bool {
-    if let hir::ExprKind::Closure(closure) = body.value.kind
-        && let [resume_ty] = closure.fn_decl.inputs
-        && let hir::TyKind::Path(hir::QPath::LangItem(hir::LangItem::ResumeTy, ..)) = resume_ty.kind
+fn is_async_closure(cx: &LateContext<'_>, body: &hir::Body<'_>) -> bool {
+    if let hir::ExprKind::Closure(desugared_outer_closure) = body.value.kind
+        && let desugared_outer_closure_body = cx.tcx.hir().body(desugared_outer_closure.body)
+        && let hir::ExprKind::Closure(desugared_inner_closure) = desugared_outer_closure_body.value.kind
+        // NB: `async || 42` is desugared into `|| |mut __context: #[lang = "ResumeTy"] | 42`; so
+        //     we need to check if the only input was that lang item.
+        && let [compiler_generated_resume_ty] = desugared_inner_closure.fn_decl.inputs
+        && let hir::TyKind::Path(hir::QPath::LangItem(hir::LangItem::ResumeTy, ..)) =
+            compiler_generated_resume_ty.kind
```

Comment:
This fixes it for me and seems to pass all tests (modulo nits; might need to fmt, etc...)
```suggestion
    if let hir::ExprKind::Closure(desugared_inner_closure) = body.value.kind
        && let desugared_inner_closure_body = cx.tcx.hir().body(desugared_inner_closure.body)
        && let Some(GeneratorKind::Async(AsyncGeneratorKind::Closure)) = desugared_inner_closure_body.generator_kind
```

I think I explained it confusingly on zulip yesterday, sorry for that. The `body: &hir::Body<'_>` that this function receives is already the **outer** closure body, so the variable name on line 65 "desugared_outer_closure" is not right. That is already the inner closure. There's no other closures after it, so trying to get another closure won't work.
This is also why the async closure `(async || 2)().await` tests failed: this function would never return true. It's trying to get one closure too much